Cannot build when using Sharp 0.33.0

### Astro Info

```block
Astro                    v4.0.2
Node                     v18.18.2
System                   Windows (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             auto-import
                         @astrojs/mdx
                         @astrojs/tailwind
                         @astrojs/sitemap
                         astro-compress
```


### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

If I upgrade the Sharp dependency to 0.33.0 for image optimization with astro:assets, I am unable to build locally (`npm run build`). Development mode still works fine.

This happens with both Astro v3.6.4 and v4.0.2.

The error logged to the console looks like:

```bash
MissingSharp: Could not find Sharp. Please install Sharp (`sharp`) manually into your project or migrate to another image service.
    at loadSharp (file:///.../dist/chunks/astro/assets-service_x_nToqfa.mjs:352:11)
    at async Object.transform (file:///.../dist/chunks/astro/assets-service_x_nToqfa.mjs:364:15)
    at async generateImageInternal (file:///.../node_modules/astro/dist/assets/build/generate.js:119:24)
    at async generateImage (file:///.../node_modules/astro/dist/assets/build/generate.js:66:28)
    at async file:///.../node_modules/p-queue/dist/index.js:118:36 {
  loc: undefined,
  title: 'Could not find Sharp.',
  hint: "See Sharp's installation instructions for more information: https://sharp.pixelplumbing.com/install. If you are not relying on `astro:assets` to optimize, transform, or process any images, you can configure a passthrough image service instead of installing Sharp. See https://docs.astro.build/en/reference/errors/missing-sharp for more information.\n" +
    '\n' +
    'See https://docs.astro.build/en/guides/images/#default-image-service for more information on how to migrate to another image service.',
  frame: undefined,
  type: 'AstroError'
}
```

**Additional notes:** It seems to be an interaction between astro:assets and astro-compress. If I even have the line `import compress from "astro-compress"` inside `astro.config.mjs` the build will fail (I don't even have to add it to the integrations arrray). However, if I remove the compress() integration, and comment out the import statement, it will build fine. THEN I can add the astro-compress import and integration back into the integrations array and it will build and compress fine. But if I then go and add an image or otherwise cause the images to need re-generated (since they aren't cached), the issue will occur again. And then you'll have to do the comment out compress and what-not all over again.

If I downgrade sharp back to 0.32.6 it builds just fine.

### What's the expected result?

The website should build and this error should not be thrown.

### Link to Minimal Reproducible Example

[Stackblitz link](https://stackblitz.com/edit/github-ehydmu-nwlevv?file=astro.config.mjs). You need to download it, run `npm install`, then `npm run build` to see the error. See the "Additional notes" section above for more details.

### Participation

- [ ] I am willing to submit a pull request for this issue.

I was getting a build failure on Cloudflare pages: `[ERROR] [vite] [vite]: Rollup failed to resolve import "sharp"` while using nodejs `18.14.1`. I found this issue, upgraded to nodejs `20.10.0` and that resolved the issue.

--- a/.changeset/twenty-stingrays-hear.md
+++ b/.changeset/twenty-stingrays-hear.md
@@ -1,0 +1,5 @@
+---
+"astro": patch
+---
+
+Pin Sharp to 0.32.6 until we can raise our semver requirements. To use the latest version of Sharp, you can add it to your project's dependencies.

--- a/examples/starlog/package.json
+++ b/examples/starlog/package.json
@@ -12,6 +12,6 @@
   "dependencies": {
     "astro": "^4.1.1",
     "sass": "^1.69.5",
-    "sharp": "0.32.6"
+    "sharp": "^0.32.6"
   }
 }

--- a/examples/starlog/package.json
+++ b/examples/starlog/package.json
@@ -12,6 +12,6 @@
   "dependencies": {
     "astro": "^4.1.1",
     "sass": "^1.69.5",
-    "sharp": "^0.32.5"
+    "sharp": "0.32.6"
   }
 }

--- a/packages/astro/package.json
+++ b/packages/astro/package.json
@@ -179,7 +179,7 @@
     "zod": "^3.22.4"
   },
   "optionalDependencies": {
-    "sharp": "0.32.6"
+    "sharp": "^0.32.6"
   },
   "devDependencies": {
     "@astrojs/check": "^0.3.1",

--- a/packages/astro/package.json
+++ b/packages/astro/package.json
@@ -179,7 +179,7 @@
     "zod": "^3.22.4"
   },
   "optionalDependencies": {
-    "sharp": "0.32.6 || ^0.33.1"
+    "sharp": "0.32.6"
   },
   "devDependencies": {
     "@astrojs/check": "^0.3.1",

--- a/packages/astro/package.json
+++ b/packages/astro/package.json
@@ -179,7 +179,7 @@
     "zod": "^3.22.4"
   },
   "optionalDependencies": {
-    "sharp": "0.32.6"
+    "sharp": "0.32.6 || ^0.33.1"
   },
   "devDependencies": {
     "@astrojs/check": "^0.3.1",

--- a/packages/astro/package.json
+++ b/packages/astro/package.json
@@ -179,7 +179,7 @@
     "zod": "^3.22.4"
   },
   "optionalDependencies": {
-    "sharp": "^0.33.1"
+    "sharp": "0.32.6"
   },
   "devDependencies": {
     "@astrojs/check": "^0.3.1",

